src/Pagination/Pagination.component.tsx

```diff
@@ -0,0 +1,268 @@
+    current?: number;
+
+    totalPages?: number;
+    totalResults?: number;
+    pageSize?: number;
+
+    showGoto?: boolean;
+    showArrows?: boolean;
+    size?: Size;
+    variant?: Variant;
+
+    prefix?: (props: RenderProps) => any;
+    suffix?: (props: RenderProps) => any;
+    children?: (props: RenderProps) => any;
+
+    onChange?: (page: number) => any;
+}
+
+const StyledPagination = styled('div')`
+    display: flex;
+    flex-direction: row;
+    justify-content: center;
+    align-items: center;
+`;
+
+const AffixWrapper = styled('div')`
+    display: inline;
+    margin: 0 1rem;
+`;
+
+interface PaginationState {
+    current: number;
+}
+
+function reducer(state: PaginationState, action: any) {
+    switch (action.type) {
+        case 'increment':
+            return {
+                ...state,
+                current: Math.min(state.current + 1, action.total),
+            };
+        case 'decrement':
+            return { ...state, current: Math.max(state.current - 1, 1) };
+        case 'setCurrent':
+            return {
+                ...state,
+                current: constrain(1, action.payload, action.total),
+            };
+        case 'totalChanged':
+            // This will make it so that if the current page is now beyond the
+            // new page total then the current page changes to the new last page.
+            // It might be that we want to revert to the first page instead.
+            return {
+                ...state,
+                current: constrain(1, state.current, action.total),
+            };
+        default:
+            throw new Error();
+    }
+}
+
+// We could improve this component by adding ways to show the total number of results,
+// allowing "Next" and "Prev" text instead of arrows, and having the Ellipses buttons
+// dropdown a Goto or decrement/increment the current page by 5, for instance.
+export const Pagination = ({
+    className,
+    prefix,
+    suffix,
+    children,
+    totalPages,
+    totalResults,
+    pageSize = 10,
+    current: controlledCurrent,
+    showGoto,
+    showArrows = true,
+    size = 'sm',
+    onChange,
+    variant = 'pager',
+    ...props
+}: PaginationProps): React.ReactElement<PaginationProps> => {
+    const [state, dispatch] = React.useReducer(reducer, {
+        current: 1,
+    });
+
+    // If the current page is controlled we want to
+    // prefer that over internal state
+    const current = controlledCurrent || state.current;
+
+    // Prefer the totalPages prop over calculating from
+    // the totalResults/pageSize
+    const total =
+        totalPages || (totalResults && Math.ceil(totalResults / pageSize)) || 1;
+
+    React.useEffect(() => {
+        dispatch({ type: 'totalChanged', total });
+    }, [total]);
+
+    React.useEffect(() => {
+        if (onChange) {
+            onChange(state.current);
+        }
+    }, [state.current]);
+
+    const pageButton = ({ page, slot }: { page: number; slot?: number }) => (
```

Comment:
It feels a little odd (maybe even anti-pattern?) to define a component within another component's render even if it is just a wrapper. Could there also be performance hits with re-renders?

Reply:
Uh, I'm not really defining a component, just a closure. It's not the same because I'm actually calling the function, not returning it as a child element.

Reply:
Yeah, you're right, it is really a closure so that's my bad. I will admit I may be overly cautious on this. I've read a few articles where this approach was said to be poor for performance, but it's not like I've done serious perf testing on it.